### PR TITLE
Consider optional attendees when finding possible meeting times. Also adds tests.

### PR DESCRIPTION
**Pull request by margaretchan**
Primary: Derrill

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,51 +14,67 @@
...
-import java.util.Set;
+import java.util.Iterator;
 import java.util.List;
-import java.util.ArrayList;
+import java.util.Set;
 
 public final class FindMeetingQuery {
 
     /** 
      * Given all known events and a new event request, find all possible time ranges to schedule
```
Comment (derrillhd) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:31]: It's great you're documenting most of the use cases your function does.

Maybe list it out like:

0. If request has duration greater than a day, then return zero slots.
1. if no required or no optional return all day.
2. If all are optional (no required), then only return possible slots for those (treating them as if they were required).
3. If has required and no optional then return slots that try to match both required
4. If has required and optional, try to match all attendees.  If none exist, only return slots to match required only.

You might want to try to organize your code around those use cases.  Or add individual line comments to your code to try to map your code to those use cases.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -28,247 +28,394 @@
 /** */
 @RunWith(JUnit4.class)
 public final class FindMeetingQueryTest {
-  private static final Collection<Event> NO_EVENTS = Collections.emptySet();
```
Comment (derrillhd) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:31]: just complaining: geez.. stupid GitHub's differ sucks.  It's showing the whole file as different..   Sigh.

**Review by derrillhd — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -92,11 +107,55 @@ public int compare (Event e1, Event e2) {
      * If there is enough time, add a new TimeRange to the possibleTimes list.
      */
     private void sufficientTime (int startTime, int endTime, long duration,
```
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:109]: Changed to hasSufficientTime.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -92,11 +107,55 @@ public int compare (Event e1, Event e2) {
...
         int freeTime = endTime - startTime;
         if (freeTime >= duration) {
             possibleTimes.add(TimeRange.fromStartEnd(startTime, endTime, inclusive));
         }
     }
 
+    /** 
+     * Find the timeOptions which all optional attendees can attend.
+     */ 
+    private Collection<TimeRange> checkOptionalAttendees (Collection<Event> events,
```
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:120]: The events List is sorted inside the helper filterAndSortEvents. However, that helper modifies the list (creating side effects) so I will change it to return a new list called sortedEvents.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,51 +14,67 @@
...
+        
+        optionalAttendeeTimes = checkOptionalAttendees(events, requiredAttendeeTimes, request);
 
-        // Sort events by start time 
-        Collections.sort(eventsList, new Comparator<Event>() {
-            public int compare (Event e1, Event e2) {
-                return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
-            }
-        });
+        return (optionalAttendeeTimes.isEmpty() ? requiredAttendeeTimes : optionalAttendeeTimes);
```
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: Removed.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,51 +14,67 @@
...
+        List<Event> eventsList = new ArrayList<>(events);
+
+        Collection<TimeRange> possibleTimes = new ArrayList<>();
+
+        filterAndSortEvents(eventsList, attendees);
 
-        if (eventsList.size() == 0) {
-            sufficientTime(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, duration, 
-                            possibleTimes, true);
+        if (eventsList.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:72]: Changed.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:120]: Changed to filterForOptionalAttendees.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:109]: Changed to have the helper return an Optional TimeRange. The caller then adds the TimeRange to possibleTimes.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -83,7 +99,6 @@ public int compare (Event e1, Event e2) {
                 sufficientTime(currEventEnd, TimeRange.END_OF_DAY, duration, possibleTimes, true);
```
[the remark this answers is not in the document]
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:99]: Do you mean /*inclusive=*/? It doesn't seem to recognize the syntax of /inclusive=/.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (margaretchan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:31]: Added detailed comments about behavior in different cases and maps to the code.

**Pull request merged**